Add Val Type

PhilipVinc opened the pull request:
This PR adds a `Val` type (similar to julia's Val type).
To do so, I add a hook to override the deduction of the type-parameters when calling `parametrictype(args...)`.

The hook is simply a function, provided to the `@parametric` decorator, that takes the arguments as input and must return the type parameters as output.
The default mechanism is now switched to a default implementation of this function.

I need to document what is added in this PR, but i'd like to know if you'd accept this first.

--

The motivation for this is that i have some bool values in netket that i'd like to use to dispatch on the right implementation.
As those values are jax-jit time constants, it makes sense to use them for dispatch to the right implementation...

wesselb commented:
As with your other PRs, also this looks great!!

Instead of adding a keyword `init_args_types` to `parametric`, how would you feel about instead checking for a magic method `__infer_type_parameter__`, in a way that the below would be the implementation of `NTuple`?

```python
from typing import Union
from plum import parametric, type_parameter

@parametric
class NTuple
    def __init__(self, *vals):
        n, val_type = type_parameter(self)
        assert all([isinstance(val, val_type) for val in vals])
        assert len(vals) == n
        self.vals = vals

    def __infer_type_parameter__(self, *vals):
        return Union[tuple(type(val) for val in vals)], len(vals)

```

If the magic method doesn't exist, then the fallback should be the current behaviour.

PhilipVinc commented:
I had thought about that at first, and indeed from one point of view it feels cleaner.

However i had thought about the following counter-argument:
 - This method should live in the metaclass, not the class itself. The first argument shouldn't be `self` but `cls`, and it should be marked as a `static` method (not even, because it's called before the class is actually created) so it felt a bit weird to put it in the class.
- if i slightly change the signature of `init_args_types` (or `infer_type_parameters`, which is much better), in order not to be `infer_type_parameters(cls, val1, val2=3)...` one can use `@dispatch` on the infer_type function itself. Maybe this is useless, but maybe it might be interesting. In any case it's more general.

Point 1 is a bit moot, so could be ignored with a bit of metaprogramming in order to extract the underlying function from the method.
Point 2 could also be ignored, because it's possible to declare dispatch rules also on class methods but i'm unsure of how it will work together with the workaround to point 1.

wesselb commented:
Those are two good points!

Regarding the first point, we could require `__infer_type_parameter__` to be a `@classmethod`. Moreover, regarding the second point, the user would be able to also use `@dispatch` within the class:

```python
from typing import Union
from numbers import Number
from plum import dispatch, parametric, type_parameter

@parametric
class NTupleForNumbers
    def __init__(self, *vals):
        n, val_type = type_parameter(self)
        assert all([isinstance(val, val_type) for val in vals])
        assert len(vals) == n
        self.vals = vals

    @dispatch
    @classmethod
    def __infer_type_parameter__(cls, *vals: Number):
        return Union[tuple(type(val) for val in vals)], len(vals)
```

(Currently, `@dispatch` appears to not work with `@staticmethod` and `@classmethod`, but that should be easy to fix.)

What do you think? It feels kinda nice to bundle `__infer_type_parameter__` together with the class and to keep `parametric` just a simple decorator.

Just letting you know that I'm currently on holiday for a few days. I'll get back to this PR and your other PRs early next week!

PhilipVinc commented:
That's a nice idea!
I implemented it (classmethods still dont work with dispatch, but that can be addressed in a separate PR).

By the way, i just noticed that maybe you already had something like `Val` type? 
Is it what `parametric.Kind` is supposed to be?

PhilipVinc commented:
bumpety bump @wesselb , unless you're still on holiday

wesselb commented on plum/parametric.py:
```diff
@@ -54,16 +56,33 @@ def __call__(cls, *args, **kw_args):
+        metaclass.
+
+        Args:
+            *args: the argument-values passed to the __init__ method.
```
```suggestion
            *args: Positional arguments passed to the `__init__` method.
            *kw_args: Keyword arguments passed to the `__init__` method.
```

wesselb commented on plum/parametric.py:
```diff
@@ -54,16 +56,33 @@ def __call__(cls, *args, **kw_args):
+            *args: the argument-values passed to the __init__ method.
+
+        Returns:
+            A type or tuple of types.
```
```suggestion
            type or tuple[type]: A type or tuple of types.
```

wesselb commented on plum/parametric.py:
````diff
@@ -102,7 +121,20 @@ def __subclasscheck__(self, subclass):
+
+    ```python
+    @classmethod
+    __infer_type_parameter__(cls, *vals, **kw_args) -> Tuple:
````
```suggestion
    __infer_type_parameter__(cls, *args, **kw_args) -> Tuple:
```

wesselb commented on plum/parametric.py:
```diff
@@ -293,3 +325,40 @@ def type_of(obj: tuple):
+
+@parametric
+class Val:
+    """
+    A parametric type used to move information from the value-domain to the type-domain.
```
```suggestion
    """A parametric type used to move information from the value domain to the type domain.
```

wesselb commented on plum/parametric.py:
```diff
@@ -293,3 +325,40 @@ def type_of(obj: tuple):
+    """
+
+    @classmethod
+    def __infer_type_parameter__(cls, *arg, **kwargs):
```
```suggestion
    def __infer_type_parameter__(cls, *arg, **kw_args):
```

wesselb commented on plum/parametric.py:
```diff
@@ -293,3 +325,40 @@ def type_of(obj: tuple):
+        if len(arg) == 0:
+            raise ValueError("The value must be specified.")
+        elif len(arg) > 1:
+            raise ValueError("Too many values. Val accepts only one argument.")
```
```suggestion
            raise ValueError("Too many values. `Val` accepts only one argument.")
```

wesselb commented on plum/parametric.py:
```diff
@@ -293,3 +325,40 @@ def type_of(obj: tuple):
+
+    def __init__(self, arg=None):
+        """
+        Construct a Value object with type `Val(arg)` that cna be used to dispatch
```
```suggestion
        Construct a value object with type `Val(arg)` that can be used to dispatch
```

wesselb commented on plum/parametric.py:
```diff
@@ -293,3 +325,40 @@ def type_of(obj: tuple):
+        based on values.
+
+        Args:
+            arg: the value to be moved to the type domain
```
```suggestion
            arg (object): The value to be moved to the type domain.
```

wesselb commented on plum/parametric.py:
```diff
@@ -293,3 +325,40 @@ def type_of(obj: tuple):
+        if type(self).is_concrete:
+            return
+        else:
+            raise ValueError("The value must be specified.")  # pragma: no cover
```
```suggestion
            raise ValueError("The value must be specified.")
```

(I think this is tested.)

wesselb commented on plum/parametric.py:
```diff
@@ -293,3 +325,40 @@ def type_of(obj: tuple):
+            raise ValueError("The value must be specified.")  # pragma: no cover
+
+    def __repr__(self):
+        return "{}()".format(repr(type(self)))
```
```suggestion
        return f"{repr(type(self))}()"
```

wesselb commented on plum/parametric.py:
```diff
@@ -293,3 +325,40 @@ def type_of(obj: tuple):
+    def __repr__(self):
+        return "{}()".format(repr(type(self)))
+
+    def __eq__(self, other):
```
```suggestion
    @_dispatch
    def __eq__(self, other: "Val"):
```

Maybe add this for safety? Perhaps also add a quick test to test this?

PhilipVinc replied:
Why? 
I think this is already pretty safe, and avoids an extra dispatch...
Unless we want to make this extensible (but do we?)

wesselb replied:
Hmm, yeah, I was worried about the case where `other` isn't a `Val`, but checking the types should catch that.

PhilipVinc replied:
What do you mean by checking the types?

wesselb replied:
The implementation  of `__eq__` is `type(self) == type(other)`, so that can never be true if `other` isn't a `Val`, because then the type of `other` will be different. Think this looks good. :)

wesselb commented on tests/test_parametric.py:
```diff
@@ -121,6 +122,24 @@ def __init__(self, x, y):
+    @parametric
+    class NTuple:
+        @classmethod
+        def __infer_type_parameter__(self, *vals, **kwargs):
+            # mimicks the type parameters of an NTuple
```
```suggestion
        def __infer_type_parameter__(self, *vals, **kw_args):
            # Mimicks the type parameters of an `NTuple`.
```

wesselb commented on tests/test_parametric.py:
```diff
@@ -302,3 +321,25 @@ def f(x: NPArray[2]):
         f(np.random.randn(10, 10, 10))
+
+
+def test_Val():
```
```suggestion
def test_val():
```

wesselb commented on plum/parametric.py:
```diff
@@ -54,16 +56,33 @@ def __call__(cls, *args, **kw_args):
+        Returns:
+            A type or tuple of types.
+        """
+        type_parameter = tuple(type(arg) for arg in args)
```
Should we consider using `type_of` here?

PhilipVinc replied:
fine by me

wesselb reviewed (changes requested):
Hey! I've been back for a few days. Thanks for the bump!

I really like this pattern with `__infer_type_parameter__`. This looks great.

I've left some minor comments.

wesselb commented:
Also, would you also like to add an entry to the README? I'm also happy to do that, if you like.

PhilipVinc commented:
I added to the readme.

wrt using `type_of`: i reverted the commit.
It turns out that problems arise in the tests because you cannot really compare plum types with standard python types:

```python
>>> plum.type_of(float) == float
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/filippovicentini/Dropbox/Ricerca/Codes/Python/plum/plum/util.py", line 30, in __eq__
    return self <= other <= self
  File "/home/filippovicentini/Dropbox/Ricerca/Codes/Python/plum/plum/type.py", line 103, in __le__
    
TypeError: issubclass() arg 1 must be a class
```

I think this PR could be merged as-is, and eventually this can be left as an improvement for the future if this issue is resolved

wesselb commented:
> I think this PR could be merged as-is, and eventually this can be left as an improvement for the future if this issue is resolved

Ah, that's unfortunate. I agree: let's leave it for a future issue.

Think this is ready to be merged then! As many times before, thanks @PhilipVinc for putting together another great PR :)

wesselb commented:
This is included in the latest release `v1.4.0`. Thanks again :)
